File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int preI = 0;
    int inI = 0;

    TreeNode* dfs(vector<int>& preorder, vector<int>& inorder, int split){
        if(preI >= preorder.size()) return nullptr;
        if(inorder[inI] == split){
            inI++;
            return nullptr;
        }

        TreeNode* root = new TreeNode(preorder[preI++]);
        root->left = dfs(preorder, inorder, root->val);
        root->right = dfs(preorder, inorder, split);

        return root;
    }

    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return dfs(preorder, inorder, INT_MAX); 
    }
};

File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Codec:
    
    # Encodes a tree to a single string.
    def serialize(self, root: Optional[TreeNode]) -> str:
        rslt = []
        def traverse(node):
            if node is None:
                rslt.append('N')
                return
            
            rslt.append(str(node.val))
            traverse(node.left)
            traverse(node.right)
        
        traverse(root)
        return ','.join(rslt)
            
    # Decodes your encoded data to tree.
    def deserialize(self, data: str) -> Optional[TreeNode]:
        vals = iter(data.split(','))

        def traverse():
            v = next(vals)
            if v == 'N':
                return None
            
            node = TreeNode(int(v), None, None)
            node.left = traverse()
            node.right = traverse()
            return node

        return traverse()